Reland "AOTAutograd: Go down inference path if no outputs require grad (#111011)"

bdhirsh opened the pull request:
Re-land of https://github.com/pytorch/pytorch/pull/111011.

The original PR ended up having a bad interaction with code that tried to run `torch.compile` under `with torch.inference_mode`, which caused some internal tests to fail.

The issue was that:

(1) AOTInductor invokes the pattern matcher passes in inductor

(2) The pattern matcher registers some code with [training_graph](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L461)

(3) The `training_graph` function expects to be able to set the global autograd state to `requires_grad`, and always get out a join graph (assertion [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L1196)).

(4) However, when inference_mode is activated, and you try to run AOTAutograd, AOTAutograd will witness that all outputs to the traced function will not require grad, and (now correctly) think that we are tracing an inference graph, which fails the above assert.

After talking to Bin, it sounds like these training-only patterns aren't necessary when we know we are compiling an inference graph (which should always be the case if you're running torch.compile with inference_mode). So I updated the pattern matcher to ignore any pattern matches using `training_graph`, when inference_mode is enabled.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111347

This reverts commit cf6b1cdf6ac74d375b0787bd8f9463cb3a53b0e5.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

bdhirsh committed:
Reland "AOTAutograd: Go down inference path if no outputs require grad (#111011)"

This reverts commit cf6b1cdf6ac74d375b0787bd8f9463cb3a53b0e5.

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -907,6 +907,12 @@ def normalize_args(**kwargs):
+        # If inference mode is enabled during compilation, assume that we don't
+        # want to match on any training graph patterns
+        if torch.is_inference_mode_enabled():
+            return False
```
cc @Chillee. More details in the PR description, but let me know what you think.

Chillee replied:
Hmm... this is a bit weird to me, but I guess it's fine.

In particular, there's no reason an "inference" graph can't also have joint graph patterns apply to it. For example, if we capture forwards and backwards in a single graph, we might go down the inference graph path anyways?

I guess this wouldn't work today anyways, as these passes are hardcoded to trigger in the "joint graph" part. But something to keep in mind in the future.

bdhirsh committed:
Update on "Reland "AOTAutograd: Go down inference path if no outputs require grad (#111011)""

Re-land of https://github.com/pytorch/pytorch/pull/111011.

The original PR ended up having a bad interaction with code that tried to run `torch.compile` under `with torch.inference_mode`, which caused some internal tests to fail.

The issue was that:

(1) AOTInductor invokes the pattern matcher passes in inductor

(2) The pattern matcher registers some code with [training_graph](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L461)

(3) The `training_graph` function expects to be able to set the global autograd state to `requires_grad`, and always get out a join graph (assertion [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L1196)).

(4) However, when inference_mode is activated, and you try to run AOTAutograd, AOTAutograd will witness that all outputs to the traced function will not require grad, and (now correctly) think that we are tracing an inference graph, which fails the above assert.

After talking to Bin, it sounds like these training-only patterns aren't necessary when we know we are compiling an inference graph (which should always be the case if you're running torch.compile with inference_mode). So I updated the pattern matcher to ignore any pattern matches using `training_graph`, when inference_mode is enabled.


This reverts commit cf6b1cdf6ac74d375b0787bd8f9463cb3a53b0e5.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot help

bdhirsh commented:
@pytorchbot --help

bdhirsh commented:
@pytorchbot merge -i
